packages/ui/src/hooks/entities.ts

```diff
@@ -20,128 +24,53 @@ export interface EntitiesContextResult {
 
 export const useEntities = (): EntitiesContextResult => {
   const [sourceCode, setSourceCode] = useState<string>('');
-  const [entities, setEntities] = useState<BaseCamelEntity[]>([]);
-  const [visualEntities, setVisualEntities] = useState<BaseVisualCamelEntity[]>([]);
   const eventNotifier = useMemo(() => new EventNotifier(), []);
-  const [currentSchemaType, setCurrentSchemaType] = useState<SourceSchemaType>(SourceSchemaType.Route);
+  const [camelResource, setCamelResource] = useState<CamelResource>(createDefaultCamelResource());
 
   /** Set the Source Code and updates the Entities */
   const setCode = useCallback((code: string) => {
     try {
       setSourceCode(code);
       const result = parse(code);
-      const rawEntities = Array.isArray(result) ? result : [result];
-      setCurrentSchemaType(getEntityType(rawEntities[0]));
-
-      const entitiesHolder = rawEntities.reduce(
-        (acc, rawEntity) => {
-          const entity = getEntity(rawEntity);
-
-          if (entity instanceof CamelRoute || entity instanceof KameletBinding || entity instanceof Pipe) {
-            acc.visualEntities.push(entity);
-          } else if (isDefined(entity) && typeof entity === 'object') {
-            acc.entities.push(entity);
-          }
-
-          return acc;
-        },
-        { entities: [], visualEntities: [] } as {
-          entities: BaseCamelEntity[];
-          visualEntities: BaseVisualCamelEntity[];
-        },
-      );
-
-      setEntities(entitiesHolder.entities);
-      setVisualEntities(entitiesHolder.visualEntities);
+      const camelResource = createCamelResource(result);
+      setCamelResource(camelResource);
 
       /** Notify subscribers that a `entities:update` happened */
       eventNotifier.next('entities:update', undefined);
     } catch (e) {
-      setEntities([]);
-      setVisualEntities([]);
+      setCamelResource(createDefaultCamelResource());
       console.error(e);
     }
   }, []);
 
   /** Updates the Source Code whenever the entities are updated */
   const updateCodeFromEntities = useCallback(() => {
-    let code = '';
-    if (visualEntities.length == 1 && !(visualEntities[0] instanceof CamelRoute)) {
-      // Camel K CRs
-      // TODO non-visual entities have to be combined into CR
-      code = stringify(visualEntities[0], null) + '\n';
-    } else {
-      if (visualEntities.length > 0) {
-        code += stringify(visualEntities, null, { indent: 2 }) + '\n';
-      }
-      if (entities.length > 0) {
-        code += stringify(entities, null, { indent: 2 }) + '\n';
-      }
-    }
-    /** Set the Source Code directly, without using `setCode` as updating the entities is already done */
+    const code = stringify(camelResource.toJSON()) || '';
```

Comment:
[minor] If the object implements a `toJSON` method, it will be used automatically by the `YAML.stringify` and `JSON.stringify` functions.
```suggestion
    const code = stringify(camelResource) || '';
```

Reply:
done :+1: